ocaml/xapi/xapi_host.ml

```diff
@@ -596,48 +610,91 @@ let evacuate ~__context ~host ~network =
           ()
     )
     plans ;
-  (* Do it *)
-  let individual_progress = 1.0 /. float (Hashtbl.length plans) in
-  let migrate_vm vm plan =
+
+  let migrate_vm ~rpc ~session_id (vm, plan) =
+    let options = [("live", "true")] in
     match plan with
+    | Migrate host when network = Ref.null ->
+        Client.Client.Async.VM.pool_migrate ~rpc ~session_id ~vm ~host ~options
     | Migrate host ->
-        ( try
-            ( if network <> Ref.null then
-                let hosts = Db.Host.get_all ~__context in
-                List.iter
-                  (fun host ->
-                    ignore
-                    @@ Xapi_network_attach_helpers
-                       .assert_valid_ip_configuration_on_network_for_host
-                         ~__context ~self:network ~host
-                  )
-                  hosts
-            ) ;
-            let with_network_option =
-              if network <> Ref.null then
-                [("network", Ref.string_of network)]
-              else
-                []
-            in
-            let options = ("live", "true") :: with_network_option in
-            Helpers.call_api_functions ~__context (fun rpc session_id ->
-                Client.Client.VM.pool_migrate ~rpc ~session_id ~vm ~host
-                  ~options
-            )
-          with
-        | Api_errors.Server_error (code, _)
-          when code = Api_errors.vm_bad_power_state ->
-            ()
-        | e ->
-            raise e
-        ) ;
-        let progress = Db.Task.get_progress ~__context ~self:task in
-        TaskHelper.set_progress ~__context (progress +. individual_progress)
+        let options = ("network", Ref.string_of network) :: options in
+        Db.Host.get_all ~__context
```

Comment:
Hmm... since this is checking all hosts and nothing specific to the VM being migrated, it seems unnecessary to do this for each migrate call. We should be able to do it once, in the beginning of the evacuate function?

Reply:
The options are the same for each VM as well.

Reply:
We would have to collect all the hosts that we see in the plans and check these once at the beginning.

Reply:
Could to that as well, but the current code actually checks _all_ hosts and doesn't look at the plan.